### PR TITLE
MAINT: make Py_SET_SIZE and Py_SET_TYPE macros a bit safer

### PR DESCRIPTION
**Pull request by tacaswell**
This is the spelling of the compatibility macro that CPython is
recommending.

See https://github.com/python/cpython/pull/20610

### CODE REVIEW

**Comment by eric-wieser** [comment]
This looks less safe to me, because now it can't be used in an expression like `(Py_SET_TYPE(x, y), z)`, which is legal in 3.10.

**Comment by tacaswell** [comment]
I am going to back out of this discussion as I am out of my depth on c/c++ details and let you and Victor sort this out :)

**Comment by seberg** [comment]
Would it be OK to make it a milestones issue rather?

File: numpy/core/include/numpy/npy_3kcompat.h [hunk 1]
```diff
@@ -62,9 +62,9 @@ static NPY_INLINE int PyInt_Check(PyObject *op) {
 
 #if PY_VERSION_HEX < 0x030900a4
     /* Introduced in https://github.com/python/cpython/commit/d2ec81a8c99796b51fb8c49b77a7fe369863226f */
-    #define Py_SET_TYPE(obj, typ) (Py_TYPE(obj) = typ)
+    #define Py_SET_TYPE(obj, type) do { Py_TYPE(obj) = (type); } while (0)
     /* Introduced in https://github.com/python/cpython/commit/b10dc3e7a11fcdb97e285882eba6da92594f90f9 */
-    #define Py_SET_SIZE(obj, size) (Py_SIZE(obj) = size)
+    #define Py_SET_SIZE(obj, size) do { Py_SIZE(obj) = (size); } while (0)
```
Comment (eric-wieser) [inline numpy/core/include/numpy/npy_3kcompat.h:67]:
```suggestion
    #define Py_SET_SIZE(obj, size) ((Py_SIZE(obj) = (size)), (void)0)
```

**Review by eric-wieser — COMMENTED** [review]
From the other PR

**Comment by eric-wieser** [comment]
@seberg, this is valuable as an open PR because it means it's super easy to edit in-browser :)

**Comment by vstinner** [comment]
> This looks less safe to me, because now it can't be used in an expression like (Py_SET_TYPE(x, y), z), which is legal in 3.10.

Do you mean that https://github.com/python/cpython/pull/20610 should suggest to use `#define Py_SET_SIZE(obj, size) (Py_SIZE(obj) = (size)) `?

**Comment by eric-wieser** [comment]
The variant I suggested over there using `(void)0` is better than both the original here and the original there. Our version was unsafe because it allowed `Py_SET_TYPE(x, y) = z`, which obviously won't work in 3.9. Your version was overly strict. The version that is now in both patches should be exactly the right level of strict :)

**Comment by seberg** [comment]
Did not expect as fast turn-around... I am half expecting that in the end, there may be some python header  we can vendor in here instead. But for now all good as soon as tests pass...

I don't mind these small changes there are easy enough, just if there is a mass of tiny macro fixes we backport one-by-one it will get annoying.

**Comment by vstinner** [comment]
> Did not expect as fast turn-around... I am half expecting that in the end, there may be some python header we can vendor in here instead.

I started a thread on the capi-sig list to propose exactly that:
https://mail.python.org/archives/list/capi-sig@python.org/thread/HUCHXBYCHL6ZXPZYOEEATXTKMQLAYVEA/

**Pull request merged**